src/persist/src/operators/stream.rs

```diff
@@ -41,46 +50,215 @@ where
         Stream<G, (String, u64, isize)>,
     ) {
-        let (write, read) = token;
-
-        let (ok_new, err_new) = self.ok_err(
-            move |((k, v), ts, diff)| -> Result<((K, V), u64, isize), (String, u64, isize)> {
-                let res = write.write(&[((k.clone(), v.clone()), ts, diff)]);
-                // TODO: Don't do this synchronously.
-                res.recv().map_err(|err| {
-                    let err_str = format!("persisting data: {}", err);
-                    (err_str, ts, 1)
-                })?;
-                Ok(((k, v), ts, diff))
-            },
-        );
-
-        // Replay the previously persisted data, if any.
-        let (ok_previous, err_previous) = match read.snapshot() {
-            Err(err) => (
-                operator::empty(&self.scope()),
-                // TODO: Figure out how to make these retractable.
-                vec![(format!("replaying persisted data: {}", err), 0, 1)]
-                    .to_stream(&mut self.scope()),
-            ),
-            Ok(snapshot) => operators::replay(&mut self.scope(), snapshot),
-        };
-
-        let ok_all = ok_previous.concat(&ok_new);
-        let err_all = err_previous.concat(&err_new);
-        (ok_all, err_all)
+        let operator_name = format!("seal({})", name);
+        let mut seal_op = OperatorBuilder::new(operator_name.clone(), self.scope());
+
+        let mut data_input = seal_op.new_input(&self, Pipeline);
+
+        let (mut data_output, data_output_stream) = seal_op.new_output();
+        let (mut error_output, error_output_stream) = seal_op.new_output();
+
+        let mut data_buffer = Vec::new();
+        let mut input_frontier =
+            Antichain::from_elem(<G::Timestamp as timely::progress::Timestamp>::minimum());
+
+        let active_seal_operator = self.scope().index() == 0;
+
+        seal_op.build(move |mut capabilities| {
+            let mut err_cap = Some(capabilities.pop().expect("missing capability"));
+            let mut _data_cap = capabilities.pop();
+
+            move |frontiers| {
+                let mut data_output = data_output.activate();
+                let mut error_output = error_output.activate();
+
+                // pass through data
+                data_input.for_each(|time, data| {
+                    data.swap(&mut data_buffer);
+
+                    let mut session = data_output.session(&time);
+                    session.give_vec(&mut data_buffer);
+                });
+
+                // seal if the frontier advances
+                let new_input_frontier = frontiers[0].frontier();
+                let progress =
+                    !PartialOrder::less_equal(&new_input_frontier, &input_frontier.borrow());
+
+                if progress {
+                    input_frontier.clear();
+                    input_frontier.extend(new_input_frontier.into_iter().cloned());
+
+                    if let Some(data_frontier) = input_frontier.elements().first() {
```

Comment:
ditto hardcoding single element antichains

Reply:
Agreed! Should I change this code to iterate through all elements and seal? Or should we change `seal()` to take an `Antichain<u64>` instead of a single `u64`. I know that internally, `seal()` does a `Antichain::from_elem(ts)` right now.

Reply:
I'm fine with making seal take an Antichain<u64>. Happy to pick that off and let you rebase on top of it if you don't want to deal with it

Reply:
Sure, I guess it will be quicker when you do it. :ok_hand:

Reply:
One thing to note, though. There is an asymmetry in timely itself, where input frontiers are antichains, but capabilities are usually just for a timestamp. So an operator would, say, have `Antichain<u64>` as input frontiers but hold on to `Capability<u64>`, where `capability.downgrade(&u64)` roughly corresponds to `seal(&u64)`.

Reply:
Okay that's interesting. The closest analog to `seal` that exists in timely/differential is the way dd's Arrange operator moves things from its internal buffer into a Trace, so lets see how that works...

It looks like that happens by adding a Batch to the trace where the Batch's lower is required to match the current upper of the trace. In practice, I think this batch's upper antichain only ever differs from its lower by a single element, but it looks like this is a bit of awkwardness due to exactly the timely asymmetry you mention:

- https://docs.rs/differential-dataflow/0.12.0/src/differential_dataflow/operators/arrange/arrangement.rs.html#586-589
- https://docs.rs/differential-dataflow/0.12.0/src/differential_dataflow/operators/arrange/arrangement.rs.html#602-605

I haven't thought through all the implications of this, but given that you're going to have to advance the frontier/drop capabilities after the seal succeeds, maybe it is more natural to iterate over the elements and individually seal the ones that have advanced?

Reply:
I think so as well. I pushed a commit that does that.